src/Func.cpp

```diff
@@ -602,529 +608,424 @@ class SubstituteSelfReference : public IRMutator {
- * Any reference to the RVar in the predicates will be replaced with reference
- * to a Var. */
-bool apply_purify(const Split &s, vector<ReductionVariable> &rvars,
-                  vector<Expr> &predicates, vector<Expr> &args,
-                  vector<Expr> &values, map<string, Expr> &dim_extent_alignment) {
-    internal_assert(s.split_type == Split::PurifyRVar);
-    const auto &iter = std::find_if(rvars.begin(), rvars.end(),
-                                    [&s](const ReductionVariable &rv) { return (s.old_var == rv.var); });
-    if (iter != rvars.end()) {
-        debug(4) << "  Purify RVar " << iter->var << " into Var " << s.outer
-                 << ", deleting it from the rvars list\n";
-        rvars.erase(iter);
+    return std::nullopt;
+}
 
-        vector<ApplySplitResult> splits_result = apply_split(s, "", dim_extent_alignment);
-        vector<pair<string, Expr>> bounds_let_stmts = compute_loop_bounds_after_split(s, "");
-        apply_split_result(bounds_let_stmts, splits_result, predicates, args, values);
+using SubstitutionMap = std::map<string, Expr>;
 
-        return true;
+// This is a helper function for building up a substitution map that
+// corresponds to pushing down a nest of lets. The lets should be fed
+// to this function from innermost to outermost. This is equivalent to
+// building a let-nest as a one-hole context and then simplifying.
+void add_let(SubstitutionMap &proj, const string &name, const Expr &value) {
+    for (auto &[_, e] : proj) {
+        e = substitute(name, value, e);
     }
-    return false;
+    proj.emplace(name, value);
 }
 
-/** Apply rename directives on the reduction variables. */
-bool apply_rename(const Split &s, vector<ReductionVariable> &rvars,
-                  vector<Expr> &predicates, vector<Expr> &args,
-                  vector<Expr> &values, map<string, Expr> &dim_extent_alignment) {
-    internal_assert(s.split_type == Split::RenameVar);
-    const auto &iter = std::find_if(rvars.begin(), rvars.end(),
-                                    [&s](const ReductionVariable &rv) { return (s.old_var == rv.var); });
-    if (iter != rvars.end()) {
-        debug(4) << "  Renaming " << iter->var << " into " << s.outer << "\n";
-        iter->var = s.outer;
+pair<ReductionDomain, SubstitutionMap> project_rdom(const vector<Dim> &dims, const Definition &def) {
+    const auto &rdom = def.schedule().rvars();
+    const auto &splits = def.schedule().splits();
+    const auto &predicate = def.predicate();
 
-        vector<ApplySplitResult> splits_result = apply_split(s, "", dim_extent_alignment);
-        vector<pair<string, Expr>> bounds_let_stmts = compute_loop_bounds_after_split(s, "");
-        apply_split_result(bounds_let_stmts, splits_result, predicates, args, values);
-
-        return true;
+    // Compute new bounds for the RDom by walking backwards through the splits.
+    SubstitutionMap bounds_projection{};
```

Comment:
A comment explaining what you mean by "projection" in this context is probably a good idea.

Reply:
Added!